stackslib/src/chainstate/stacks/auth.rs

```diff
@@ -314,6 +315,204 @@ impl MultisigSpendingCondition {
+        }
+
+        Ok(OrderIndependentMultisigSpendingCondition {
+            signer,
+            nonce,
+            tx_fee,
+            hash_mode,
+            fields,
+            signatures_required,
+        })
+    }
+}
+
+impl OrderIndependentMultisigSpendingCondition {
+    pub fn push_signature(
+        &mut self,
+        key_encoding: TransactionPublicKeyEncoding,
+        signature: MessageSignature,
+    ) -> () {
+        self.fields
+            .push(TransactionAuthField::Signature(key_encoding, signature));
+    }
+
+    pub fn push_public_key(&mut self, public_key: StacksPublicKey) -> () {
+        self.fields
+            .push(TransactionAuthField::PublicKey(public_key));
+    }
+
+    pub fn pop_auth_field(&mut self) -> Option<TransactionAuthField> {
+        self.fields.pop()
+    }
+
+    pub fn address_mainnet(&self) -> StacksAddress {
+        StacksAddress {
+            version: C32_ADDRESS_VERSION_MAINNET_MULTISIG,
+            bytes: self.signer.clone(),
+        }
+    }
+
+    pub fn address_testnet(&self) -> StacksAddress {
+        StacksAddress {
+            version: C32_ADDRESS_VERSION_TESTNET_MULTISIG,
+            bytes: self.signer.clone(),
+        }
+    }
+
+    /// Authenticate a spending condition against an initial sighash.
+    /// In doing so, recover all public keys and verify that they hash to the signer
+    /// via the given hash mode.
+    pub fn verify(
+        &self,
+        initial_sighash: &Txid,
+        cond_code: &TransactionAuthFlags,
+    ) -> Result<Txid, net_error> {
+        let mut pubkeys = vec![];
+        let mut num_sigs: u16 = 0;
+        let mut have_uncompressed = false;
+        for field in self.fields.iter() {
+            let pubkey = match field {
+                TransactionAuthField::PublicKey(ref pubkey) => {
+                    if !pubkey.compressed() {
+                        have_uncompressed = true;
+                    }
+                    pubkey.clone()
+                }
+                TransactionAuthField::Signature(ref pubkey_encoding, ref sigbuf) => {
+                    if *pubkey_encoding == TransactionPublicKeyEncoding::Uncompressed {
+                        have_uncompressed = true;
+                    }
+
+                    let (pubkey, _next_sighash) = TransactionSpendingCondition::next_verification(
+                        &initial_sighash,
+                        cond_code,
+                        self.tx_fee,
+                        self.nonce,
+                        pubkey_encoding,
+                        sigbuf,
+                    )?;
+                    num_sigs = num_sigs
+                        .checked_add(1)
+                        .ok_or(net_error::VerifyingError("Too many signatures".to_string()))?;
+                    pubkey
+                }
+            };
+            pubkeys.push(pubkey);
+        }
+
+        if num_sigs < self.signatures_required {
```

Comment:
While I like the idea of supporting more than the required number of signatures, I'm not sure this code is handling the case where there are a surplus number of signatures.  In particular, if there are more than the minimum number of signatures, then at least one signature would need to decode to a public key that's already present in `fields`, right?  If so, then this check appears to be absent -- the extra signatures could be invalid but verification would pass.  I don't think we want this.

That is, if there are `m` signatures required, and there are `n - m` public keys given, but `m + k` signatures provided, then exactly `k` distinct signatures must recover to `k` distinct public keys out of the `n - m` public keys given.  Also, what happens if `m + k > n` -- i.e. you recover more than the `n` public keys that went into the address hash bytes?

Reply:
If it simplifies things, we could simply require _exactly_ `m` signatures like we do in the other multisig spending conditions.  That's at least consistent with the multisig experience in other blockchains.

Reply:
I've added some tests for this (It's a PR to Vlad's branch, but will be merged soon), and it works as expected:
 
| Condition                                | Valid? |
|------------------------------------------|--------|
| Extra sigs from allowed signers          | ✓      |
| Extra sigs from random signers           | ✗      |
| Number of signatures below threshold     | ✗      |

Reply:
merged @jbencin @jcnelson